bus-mapping/src/circuit_input_builder.rs

```diff
@@ -277,55 +336,101 @@ impl Call {
 }
 
 /// Context of a [`Call`].
-#[derive(Debug)]
+#[derive(Debug, Default)]
 pub struct CallContext {
     /// State Write Counter tracks the count of state write operations in the
-    /// call.  When a subcall in this call succeeds, the `swc` increases by the
+    /// call. When a subcall in this call succeeds, the `swc` increases by the
     /// number of successful state writes in the subcall.
     pub swc: usize,
+    /// State Write Counter offset of caller.
+    swc_offset: usize,
+    /// Index of Transaction.steps that triggers this call
+    step_index: usize,
 }
 
 #[derive(Debug)]
 /// Context of a [`Transaction`] which can mutate in an [`ExecStep`].
 pub struct TransactionContext {
-    /// Call Stack by indices with CallContext.
+    id: usize,
+    is_last_tx: bool,
+    /// List of calls by creation order.  Mirror of `Transaction.calls`, but to
+    /// store `CallContext`.
+    calls: Vec<CallContext>,
+    /// Call Stack by indices.
     /// The call_stack will always have a fixed element at index 0 which
     /// corresponds to the call implicitly created by the transaction.
-    call_stack: Vec<(usize, CallContext)>,
+    call_stack: Vec<usize>,
 }
 
 impl TransactionContext {
     /// Create a new Self.
-    pub fn new(_eth_tx: &eth_types::Transaction) -> Self {
+    pub fn new(id: usize, is_last_tx: bool, _eth_tx: &eth_types::Transaction) -> Self {
         Self {
-            call_stack: vec![(0, CallContext { swc: 0 })],
+            id,
+            is_last_tx,
+            calls: vec![CallContext::default()],
+            call_stack: vec![0],
         }
     }
 
+    /// Return the id of current transaction.
+    pub fn id(&self) -> usize {
+        self.id
+    }
+
+    /// Return if this transaction is the last one in block or not.
+    pub fn is_last_tx(&self) -> bool {
+        self.is_last_tx
+    }
+
     /// Return the index and context of the current call (the last call in the
     /// call stack).
     fn call_index(&self) -> usize {
-        let (index, _) = self.call_stack.last().expect("call_stack is empty");
+        let index = self.call_stack.last().expect("call_stack is empty");
```

Comment:
Yeah, invalid geth trace might still be to cause this to panic. I try with returning `Result`, otherwise we need to call `ok_or` a lot in other places. Is this a acceptable solution?